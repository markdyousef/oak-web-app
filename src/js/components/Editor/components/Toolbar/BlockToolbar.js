// @flow
import React, { PropTypes } from 'react';
import { RichUtils } from 'draft-js';
import StyleButton from '../StyleButton';
import h1Icon from '../../icons/h1';
import h2Icon from '../../icons/h2';
import qouteIcon from '../../icons/quote';
import listIcon from '../../icons/unorderedList';
import codeIcon from '../../icons/codeBlock';

const BLOCK_TYPES = [
  { label: 'H1', style: 'header-one', Icon: h1Icon },
  { label: 'H2', style: 'header-two', Icon: h2Icon },
  { label: 'Quote', style: 'blockquote', Icon: qouteIcon },
  { label: 'Bulleted List', style: 'unordered-list-item', Icon: listIcon },
  { label: 'Numbered List', style: 'ordered-list-item', Icon: listIcon },
  { label: 'Code Block', style: 'code-block', Icon: codeIcon }
];

const BlockToolbar = ({ onToggle, editorState }:Object) => {
    const blockType = RichUtils.getCurrentBlockType(editorState);
    return (
        <div>
            {BLOCK_TYPES.map(item =>
                <StyleButton
                    key={item.label}
                    active={item.style === blockType}
                    label={item.label}
                    onToggle={onToggle}
                    style={item.style}
                    Icon={item.Icon}
                />
            )}
        </div>
    );
};

BlockToolbar.propTypes = {
    editorState: PropTypes.shape({
        _immutable: PropTypes.object
    }).isRequired,
    onToggle: PropTypes.func.isRequired
};

export default BlockToolbar;
