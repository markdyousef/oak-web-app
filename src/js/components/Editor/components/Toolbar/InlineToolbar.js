// @flow
import React, { PropTypes } from 'react';
import StyleButton from '../StyleButton';
import boldIcon from '../../icons/bold';
import italicIcon from '../../icons/italic';
import underlineIcon from '../../icons/underline';

const INLINE_BUTTONS = [
    {
        label: 'B',
        style: 'BOLD',
        Icon: boldIcon,
        description: 'Bold'
    },
    {
        label: 'I',
        style: 'ITALIC',
        Icon: italicIcon,
        description: 'Italic'
    },
    {
        label: 'U',
        style: 'UNDERLINE',
        Icon: underlineIcon,
        description: 'Underline'
    },
    {
        label: 'Hi',
        style: 'HIGHLIGHT',
        Icon: underlineIcon,
        description: 'Highlight selection'
    }
];

const InlineToolbar = ({ onToggle, editorState }:Object) => {
    const currentStyle = editorState.getCurrentInlineStyle();
    return (
        <div>
            {INLINE_BUTTONS.map(item =>
                <StyleButton
                    key={item.label}
                    active={currentStyle.has(item.style)}
                    label={item.label}
                    onToggle={onToggle}
                    style={item.style}
                    Icon={item.Icon}
                />
            )}
        </div>
    );
};

InlineToolbar.propTypes = {
    editorState: PropTypes.shape({
        _immutable: PropTypes.object
    }).isRequired,
    onToggle: PropTypes.func.isRequired
};

export default InlineToolbar;
