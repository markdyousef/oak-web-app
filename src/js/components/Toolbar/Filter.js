import React, { PropTypes, Component } from 'react';

import css from './Toolbar.css';

class Filter extends Component {
    static propTypes = {
        name: PropTypes.string.isRequired,
        options: PropTypes.arrayOf(PropTypes.shape({
            name: PropTypes.string.isRequired
        })).isRequired,
        select: PropTypes.func.isRequired
    };
    constructor() {
        super();
        this.renderOptionList = this.renderOptionList.bind(this);
        this.state = {
            selected: null,
            isOpen: false
        }
    }
    renderOptionList() {
        const { options, select } = this.props;
        return options.map(option => (
            <div
                key={option.name}
                onClick={() => { select(option.name); this.setState({ isOpen: false }); }}
            >
                <span>{option.name}</span>
            </div>
        ));
    }
    render() {
        const { selected, isOpen } = this.state;
        const { name } = this.props;
        return (
            <div className={css.filter}>
                <div
                    className={css.selected}
                    onClick={() => this.setState({ isOpen: !isOpen })}
                >
                    <span>{(selected) || name }</span>
                </div>
                <div className="filter">
                    {isOpen && <div className={css.optionList}>
                        {this.renderOptionList()}
                    </div>}
                </div>
            </div>
        );
    }
}

export default Filter;
