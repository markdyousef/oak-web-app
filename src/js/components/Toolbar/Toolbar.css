:root {
    --main-black-color: #000;
    --main-white-color: #fefeff;
    --main-orange-color: #ff9200;
    --main-red-color: #8b0000;
    --main-green-color: #005d6c;
}

.container {
    width: 600px;
    height: 40px;
    background-color: var(--main-green-color);
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
}

.filter {
    width: 150px;
}

.selected {
    height: 40px;
    background-color: var(--main-white-color);
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
}

.optionList {
    width: 100%;
    background-color: var(--main-white-color);
    position: relative;
}

.optionList div {
    padding: 10px;
    cursor: pointer;
}
