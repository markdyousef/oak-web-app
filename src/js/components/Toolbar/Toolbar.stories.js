import React from 'react';
import { storiesOf } from '@kadira/storybook';
import Toolbar from './Toolbar';

storiesOf('Toolbar', module)
    .add('empty', () => (
        <Toolbar />
    ));
