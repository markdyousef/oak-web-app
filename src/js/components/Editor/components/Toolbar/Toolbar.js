import React, { Component, PropTypes } from 'react';
import styled from 'styled-components';
import { getVisibleSelectionRect } from 'draft-js';
import BlockToolbar from './BlockToolbar';
import InlineToolbar from './InlineToolbar';
import { actionsColor } from '../../styles/colors';
import { boxLayout } from '../../styles/layouts';

// Toolbar height
// const toolbarHeight = 200;

const Container = styled.div`
    width: 200px;
    ${''/* height: ${toolbarHeight}px; */}
    display: flex;
    flex-direction: column;
    ${boxLayout()};
    z-index: 2;
    transform: translate(-50%) scale(0);
    cursor: auto;
    &:after {
        content: '';
        position: absolute;
        bottom: -5px;
        left: 50%;
        margin-left: -5px;
    }
`;

const getRelativeParent = (element) => {
    if (!element) {
        return null;
    }

    const position = window.getComputedStyle(element).getPropertyValue('position');
    if (position !== 'static') {
        return element;
    }

    return getRelativeParent(element.parentElement);
};

export default class Toolbar extends Component {
    static propTypes = {
        editorState: PropTypes.shape({
            _immutable: PropTypes.object
        }),
        toggleBlockType: PropTypes.func.isRequired,
        toggleInlineStyle: PropTypes.func.isRequired,
        focus: PropTypes.func.isRequired,
        showToolbar: PropTypes.bool.isRequired,
        editorNode: PropTypes.shape({
            props: PropTypes.object
        })
    };
    static defaultProps = {
        editorState: {},
        editorNode: {}
    }
    constructor() {
        super();
        this.state = {
            position: {}
        };
    }
    componentWillReceiveProps(nextProps) {
        setTimeout(() => {
            let position;
            if (nextProps.showToolbar) {
                const relativeParent = getRelativeParent(this.toolbar.parentElement);
                const relativeRect = relativeParent ?
                    relativeParent.getBoundingClientRect()
                    :
                    document.body.getBoundingClientRect();
                position = {
                    top: (this.selectionRect.top - relativeRect.top - 30),
                    // left: (this.selectionRect.left - relativeRect.left) - (this.selectionRect.width / 2),
                    left: (this.selectionRect.left - 300),
                    transform: 'translate(-50%) scale(1)',
                    transition: 'transform 0.15s cubic-brezier(0.3, 1.2, 0.2, 1)'
                };
                // console.log('left: ' + (this.selectionRect.left - 300));
                // console.log('top: ' + (this.selectionRect.top - relativeRect.top));
                this.props.focus();
            } else {
                position = { transform: 'translate(-50%) scale(0)' };
            }
            this.setState({ position });
        });
    }
    componentDidUpdate() {
        this.selectionRect = getVisibleSelectionRect(window);
        this.props.focus();
    }
    render() {
        const { toggleBlockType, editorState, showToolbar, toggleInlineStyle } = this.props;
        const { position } = this.state;
        return (
            <Container showToolbar={showToolbar} style={position} ref={(element) => { this.toolbar = element; }}>
                <BlockToolbar
                    onToggle={toggleBlockType}
                    editorState={editorState}
                />
                <InlineToolbar
                    onToggle={toggleInlineStyle}
                    editorState={editorState}
                />
            </Container>
        );
    }
}
