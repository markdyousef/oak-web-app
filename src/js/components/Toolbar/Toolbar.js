import React, { Component, PropTypes } from 'react';
import Filter from './Filter';

import css from './Toolbar.css';

class ToolBar extends Component {
    static propTypes = {
        data: PropTypes.shape({
            filters: PropTypes.arrayOf(PropTypes.shape({
                name: PropTypes.string.isRequired,
                id: PropTypes.string,
                options: PropTypes.arrayOf(PropTypes.object)
            }))
        })
    }
    constructor() {
        super();
        this.renderFilterOptions = this.renderFilterOptions.bind(this);
        this.state = {}
    }
    renderFilterOptions() {
        const { data } = this.props;
        if (data && data.filters) {
            return data.filters.map(filter => (
                <Filter filter={filter} />
            ));
        }
    }
    render() {
        return (
            <div className={css.container}>
                {this.renderFilterOptions()}
            </div>
        );
    }
}

export default ToolBar;
