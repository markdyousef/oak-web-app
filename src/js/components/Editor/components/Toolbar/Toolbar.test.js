import React from 'react';
import { shallow } from 'enzyme';
import Toolbar from './Toolbar';

it('renders without crashing', () => {
    const props = {
        editorState: {
        },
        toggleBlockType: () => {},
        toggleInlineStyle: () => {},
        focus: () => {},
        showToolbar: true
    };
    const shallowComponent = shallow(<Toolbar {...props} />);
    expect(shallowComponent).toMatchSnapshot(shallowComponent);
});
