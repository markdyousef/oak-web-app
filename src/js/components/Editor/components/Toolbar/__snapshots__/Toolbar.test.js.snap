exports[`[object Object] 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div
    showToolbar={true}
    style={Object {}}>
    <BlockToolbar
        editorState={Object {}}
        onToggle={[Function]} />
    <InlineToolbar
        editorState={Object {}}
        onToggle={[Function]} />
</styled.div>,
  "nodes": Array [
    <styled.div
      showToolbar={true}
      style={Object {}}>
      <BlockToolbar
            editorState={Object {}}
            onToggle={[Function]} />
      <InlineToolbar
            editorState={Object {}}
            onToggle={[Function]} />
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Toolbar
        editorNode={Object {}}
        editorState={Object {}}
        focus={[Function]}
        showToolbar={true}
        toggleBlockType={[Function]}
        toggleInlineStyle={[Function]} />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Toolbar {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "editorNode": Object {},
          "editorState": Object {},
          "focus": [Function],
          "showToolbar": true,
          "toggleBlockType": [Function],
          "toggleInlineStyle": [Function],
        },
        "refs": Object {},
        "state": Object {
          "position": Object {},
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div
          showToolbar={true}
          style={Object {}}>
          <BlockToolbar
                    editorState={Object {}}
                    onToggle={[Function]} />
          <InlineToolbar
                    editorState={Object {}}
                    onToggle={[Function]} />
</styled.div>,
        "_debugID": 2,
        "_renderedOutput": <styled.div
          showToolbar={true}
          style={Object {}}>
          <BlockToolbar
                    editorState={Object {}}
                    onToggle={[Function]} />
          <InlineToolbar
                    editorState={Object {}}
                    onToggle={[Function]} />
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Toolbar
    editorNode={Object {}}
    editorState={Object {}}
    focus={[Function]}
    showToolbar={true}
    toggleBlockType={[Function]}
    toggleInlineStyle={[Function]} />,
}
`;
